### DISCUSSION TITLE
How can we serve a model for the server that has no power measurement?

### DISCUSSION THREAD

**Discussion by sunya-ch** [sustainable-computing-io/kepler-model-server#91] [Ideas]
The goal is that we want to serve a model for the server that has no power measurement.
- (i) provide a model trained by usage feature + system feature (such as CPU architecture and CPU frequency)  
- (ii) provide a model trained by usage feature trained by similar system 

For (i), we need a big data set to cover all possibilities.
For (ii), we need to determine the similarity 

For (ii), we can determine similarity by 
 - (ii-1) attribute (such as CPU info, memory info, sysctl -a)
 - (ii-2) running benchmark and use performance value such as iteration/sec of coremark 
 
 My current proposal is to use (ii-2).
 Would be great if we can discuss any idea from any method here :)

**Comment by marceloamaral** [comment]
Thanks to @sunya-ch for initiating this discussion! 

There are some typical factors that impact server consumption:
- CPU architecture (such as x86, Arm, and power)
- CPU model
- CPU frequency and voltage (they are typically linearly related though)
- CPU fan speed (some system allows changing it)
- Other components (such as GPU, storage, and NIC)
- Power supply 
- Server fan
- Server temperature (which can be affected by changes in the data center and impacts the power consumption of all internal components)
- Background workloads

While we may not have access to information on all of these factors, it is important to be as specific as possible in order to achieve good accuracy. 

This factors change the power consumption curve related to resource utilization. More specifically, a variation in this factors changes how software/hardware performance counters of resource utilization reflects the power consumption, to higher or lower values. 

> (ii-2) running benchmark and use performance value such as iteration/sec of coremark

I think that only relying on software/hardware performance counters of resource utilization does not characterize the server power consumption. **We need to know at least the CPU architecture and model**, and ideally, we should also have information on other components such as DRAM, GPU, storage, and network. 

But of course, If we have absolutely **no information on the system**, including the maximum frequency, we can just use a random power model. This is because without any information, we cannot determine the relationship between resource utilization and system power consumption, it can be anything. This is especially important for virtual machines, where the hypervisor introduces overhead and the performance may not accurately reflect the actual physical host. **But we must make sure that the user will be aware that it will be using a generic power model** that does not necessarily reflects the physical host power consumption, but reflects a given generic node.

**Comment by marceloamaral** [comment]
> (ii) provide a model trained by usage feature trained by similar system

How to define if the systems are similar? 

Determining whether systems are similar is not a simple task. This is different hardware architecture or model version may completely change the power characteristics. This includes changes to idle power, activation power, and dynamic power. Additionally, different models may have varying power capping mechanisms, which can change the power curve.

**Comment by marceloamaral** [comment]
It is possible to use a "generic" power model that has been trained with most of the provided dataset and contains all the necessary information, but only if it provides an advantage. We must verify this first, otherwise we will only waste time, resource and energy.

A particular dataset can potentially disrupt the already trained generic power model, so we must verify whether it is worthwhile to include it or not. This involves assessing the accuracy of the training dataset, all previously used datasets, and the new dataset. Therefore, we need to establish the acceptable accuracy change that we can permit by introducing a new dataset.
